### PR TITLE
home-assistant.python.pkgs.pytest-homeassistant-custom-component: init at 0.13.147

### PR DESCRIPTION
**Pull request by dotlambda**
We'd need to have Home Assistant be available as `home-assistant.python.pkgs.homeassistant`, similarly to how poetry does it. Should we do that?

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/home-assistant/pytest-homeassistant-custom-component.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  build-system = [ setuptools ];
+
+  pythonRemoveDeps = true;
+
+  dependencies = [
+    aiohttp
+    bcrypt
+    freezegun
+    pytest
```
Comment (SuperSandro2000) [inline pkgs/servers/home-assistant/pytest-homeassistant-custom-component.nix:40]: We should probably move pytest to buildInputs like in other pytest plugins.

**Comment by SuperSandro2000** [comment]
> We'd need to have Home Assistant be available as `home-assistant.python.pkgs.homeassistant`, similarly to how poetry does it. Should we do that?

Does that even work? I would suspect this triggering an infinite recursion.

**Comment by mweinelt** [comment]
> We'd need to have Home Assistant be available as home-assistant.python.pkgs.homeassistant, similarly to how poetry does it. Should we do that?

Yes.

**Comment by dotlambda** [comment]
> Does that even work? I would suspect this triggering an infinite recursion.

It doesn't if done right. Check out https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/package-management/poetry/default.nix.

**Comment by mweinelt** [comment]
Don't we just need 

> home-assistant = self.toPythonModule home-assistant;

in the package set?

**Comment by mweinelt** [comment]
Yes, that works.

```
music-assistant/mass> Executing pytestCheckPhase
music-assistant/mass> ============================= test session starts ==============================
music-assistant/mass> platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0
music-assistant/mass> rootdir: /build/source
music-assistant/mass> configfile: pyproject.toml
music-assistant/mass> plugins: homeassistant-custom-component-0.13.147, anyio-4.3.0, asyncio-0.23.6, socket-0.7.0, requests-mock-1.12.1, syrupy-4.6.1
music-assistant/mass> asyncio: mode=Mode.AUTO
music-assistant/mass> collected 7 items                                                              
music-assistant/mass> 
music-assistant/mass> tests/test_config_flow.py .......                                        [100%]
music-assistant/mass> 
music-assistant/mass> ============================== 7 passed in 0.78s ===============================
```

**Comment by dotlambda** [comment]
> ```diff
> +      home-assistant = self.toPythonModule home-assistant;
> ```

True, that also works as long as no overrides happen.

**Comment by mweinelt** [comment]
The custom components are just dropped into `/var/lib/hass/custom_components` by the module. There is no connection that I can see, that would tie them together.

**Comment by dotlambda** [comment]
> There is no connection that I can see, that would tie them together.

True, especially since custom modules are also built with top-level `home-assistant`'s Python packages.

**Comment by dotlambda** [comment]
I fail to understand why govee-lan's tests don't work:
```
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0
rootdir: /build/source
configfile: setup.cfg
testpaths: tests
plugins: homeassistant-custom-component-0.13.147, anyio-4.3.0, asyncio-0.23.6, socket-0.7.0, requests-mock-1.12.1, syrupy-4.6.1
asyncio: mode=Mode.STRICT
collected 1 item                                                               

tests/test_init.py s                                                     [100%]

=============================== warnings summary ===============================
tests/test_init.py::test_async_setup
  /nix/store/6m4rxk7rn1ncads8mfjp55hxphzyfr2j-python3.12-pytest-8.2.2/lib/python3.12/site-packages/_pytest/python.py:151: PytestUnhandledCoroutineWarning: async def functions are not natively supported and have been skipped.
  You need to install a suitable plugin for your async framework, for example:
    - anyio
    - pytest-asyncio
    - pytest-tornasync
    - pytest-trio
    - pytest-twisted
    warnings.warn(PytestUnhandledCoroutineWarning(msg.format(nodeid)))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 1 skipped, 1 warning in 0.19s =========================
```

**Comment by mweinelt** [comment]
Missing `@pytest.mark.asyncio` on the test?

**Comment by dotlambda** [comment]
I now get
```
_______________________________ test_async_setup _______________________________

hass = <async_generator object hass at 0x7ffff183aac0>

    @pytest.mark.asyncio
    async def test_async_setup(hass):
        """Test the component gets setup."""
>       assert await async_setup_component(hass, DOMAIN, {}) is True

tests/test_init.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hass = <async_generator object hass at 0x7ffff183aac0>, domain = 'govee_lan'
config = {}

    async def async_setup_component(
        hass: core.HomeAssistant, domain: str, config: ConfigType
    ) -> bool:
        """Set up a component and all its dependencies.
    
        This method is a coroutine.
        """
>       if domain in hass.config.components:
E       AttributeError: 'async_generator' object has no attribute 'config'
```
and that sounds more like the internal Home Assistant API changed.

**Pull request merged**